aten/src/THCUNN/ClassNLLCriterion.cu

```diff
@@ -125,7 +125,7 @@ __global__ void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *output,
     }
     *total_weight = ScalarConvert<Acctype, Dtype>::to(total_weightAcc);
     *output = ScalarConvert<Acctype, Dtype>::to(outputAcc);
-    if (size_average && *total_weight > 0) {
+    if (size_average && ((*total_weight != 0 && outputAcc != 0) || nframe == 0)) {
```

Comment:
Why does this expression need to be this complicated?

Reply:
It is what you asked me to fix in [previous review](https://github.com/pytorch/pytorch/pull/31491#discussion_r365461289)

Reply:
Maybe something like 
```
if (size_average) {
    if (nframe == 0) {
        *output = NaN;
    }
    if (*total_weight != 0) {
        *output = ScalarConvert<Acctype, Dtype>::to(outputAcc / total_weightAcc);
    }
}
```
makes the logic clearer